pkg/indexer/localcredentials.go

```diff
@@ -0,0 +1,103 @@
+package indexer
+
+import (
+	"context"
+	"reflect"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/apimachinery/pkg/types"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+
+	"github.com/mongodb/mongodb-atlas-kubernetes/v2/pkg/api"
+)
+
+type LocalCredentialIndexer struct {
+	obj    client.Object
+	name   string
+	logger *zap.SugaredLogger
+}
+
+// Reconciliable is implemented by CRD objects used by indexes to trigger reconciliations
+type Reconciliable interface {
+	ReconciliableRequests() []reconcile.Request
+}
+
+// ReconciliableList is a Reconciliable that is also a CRD list
+type ReconciliableList interface {
+	client.ObjectList
+	Reconciliable
+}
+
+func NewLocalCredentialsIndexer(name string, obj client.Object, logger *zap.Logger) *LocalCredentialIndexer {
+	return &LocalCredentialIndexer{
+		obj:    obj,
+		name:   name,
+		logger: logger.Named(name).Sugar(),
+	}
+}
+
+func (lc *LocalCredentialIndexer) Object() client.Object {
+	return lc.obj
+}
+
+func (lc *LocalCredentialIndexer) Name() string {
+	return lc.name
+}
+
+func (lc *LocalCredentialIndexer) Keys(object client.Object) []string {
+	if reflect.TypeOf(object) != reflect.TypeOf(lc.obj) {
+		lc.logger.Errorf("expected %T but got %T", lc.obj, object)
+		return nil
+	}
+
+	credentialProvider, ok := (object).(api.CredentialsProvider)
+	if !ok {
+		lc.logger.Errorf("expected %T to implement api.CredentialProvider", object)
+		return nil
+	}
+
+	if localRef := credentialProvider.Credentials(); localRef != nil && localRef.Name != "" {
+		return []string{types.NamespacedName{Namespace: object.GetNamespace(), Name: localRef.Name}.String()}
+	}
+	return []string{}
+}
+
+type requestsFunc[L client.ObjectList] func(L) []reconcile.Request
```

Comment:
Why is it written as a generic function when it doesn't?
```suggestion
type requestsFunc func(L client.ObjectList) []reconcile.Request
```

Reply:
Let me double check and fix accordingly, thanks!

Reply:
Actually that is required. Because `ObjectList` is not enough. When you pass the list it has to be an ObjectList, but the actual list of items is not defined in the basic Object list, but on something that extends the ObjectList.

See the first implementation:
```go
func DatabaseUserRequests(list *akov2.AtlasDatabaseUserList) []reconcile.Request {
	requests := make([]reconcile.Request, 0, len(list.Items))
	for _, item := range list.Items {
		requests = append(requests, toRequest(&item))
	}
	return requests
}
```
`AtlasDatabaseUserList` is an ObjectList, but it also needs to be able to list items, actual Atlas Database Users in this case which the ObjectList alone cannot do.